### PR TITLE
patch_parse: implement state machine for parsing patch headers

### PR DESCRIPTION
**Pull request by pks-t**
Our code parsing Git patch headers is rather lax in parsing headers of a
Git-style patch. Most notably, we do not care for the exact order in
which header lines appear and as such, we may parse patch files which
are not really valid after all. Furthermore, the state transitions
inside of the parser are not as obvious as they could be, making it
harder than required to follow its logic.

To improve upon this situation, this patch introduces a real state
machine to parse the patches. Instead of simply parsing each line
without caring for previous state and the exact ordering, we define a
set of states with their allowed transitions. This makes the patch
parser more strict in only allowing valid successions of header lines.
As the transition table is defined inside of a single structure with
the expected line, required state as well as the state that we end up
in, all state transitions are immediately obvious from just having a
look at this structure. This improves both maintainability and eases
reasoning about the patch parser.

### CODE REVIEW

**Comment by pks-t** [comment]
Rebased on current master. This is mainly to have the base same commit in #4310, which builds upon this PR.

File: src/patch_parse.c [hunk 1]
```diff
@@ -439,6 +439,7 @@ static const parse_header_transition transitions[] = {
 	/* Next patch */
 	{ "diff --git "         , STATE_END,        0,                NULL },
 	{ "@@ -"                , STATE_END,        0,                NULL },
+	{ "-- "                 , STATE_END,        0,                NULL },
```
Comment (pks-t) [inline src/patch_parse.c]: Doesn't the fix to #4322 look nice with this new state machine? :)

**Comment by pks-t** [comment]
@ethomson You think you might get some time soonish to review? I'd like to move on with the common parser interface in #4310, which depends on this PR here. More specifically, I wanted to also take a look at moving our attributes parser to use the new parsing interface, as well.

**Comment by ethomson** [comment]
THIS IS AWESOME.  🎉

**Pull request merged**